### ISSUE TITLE
Travis CI and Mac OS X support

### ISSUE THREAD

**Issue by jakirkham** [conda-forge/toolchain-feedstock#7]
While we do set our Mac OS deployment target to 10.7, it appears that Travis CI does not have the needed framework. While everything should still build correctly locally with our configuration, the binaries coming from Travis CI are not going to support that old of an OS version (at least I don't think). The reason I believe we have this problem is both due to some detective work by @isuruf in PR ( https://github.com/conda-forge/staged-recipes/pull/689 ) which shows the needed framework is missing and the info that Travis CI provides us about the [OS X build environment]( https://docs.travis-ci.com/user/osx-ci-environment/#Xcode-6.1 ) that we are using. As a result, I expect the binaries we build don't actually work on 10.7 though I could be surprised. Not having a 10.7 machine I cannot test this.

Unfortunately the oldest XCode that Travis CI provides has 10.9 support at the low end. So, we now have some tough choices to make. Do we just accept this and change the toolchain to match this constraint? While this avoids warnings and provides an accurate representation of what we support in terms of binaries, it unnecessarily constrains others who want to use our recipes to building for 10.9, which actually is not required for the functionality we wish to provide (e.g. C++11 support). In fact, 10.7 would actually work to build our recipes from source.

Alternatively, we can just install the older XCode (4.6.3). This might work, but we will still need to cache this older XCode somewhere (perhaps here) and add this installation step everywhere. Also, this will be about 1.6GBs so we might be pushing our limits on Travis CI by trying to download and install this. There is also a question of whether this older XCode will run on and behave correctly on 10.9 and how much of a pain it will be to install in Travis CI.

Another alternative would be to just try and add an old framework to Travis CI, but this is a bit tricky and I'm not sure if it will work. We could grab a framework from this [repo]( https://github.com/phracker/MacOSX-SDKs ) though frameworks have to include binary data so there is a question of trust. If we don't trust that source, we need to figure out another way to get them. We can't easily download these from Mac during CI, but we could download them locally and rip the frameworks we want out of older copies of XCode and add them perhaps to a tag here for caching. Though it is not clear to me that the binaries will behave correctly when doing this as we are basically hacking together a newer XCode with frameworks from an older XCode. In any event, this repo ( https://github.com/devernay/xcodelegacy ) provides a useful script for doing just that. Though extracting the SDK is straightforward, I'm not sure how difficult it will be to insert it correctly in a newer XCode.

No matter how we chose to go forward, we will still have C++11 support. This only affects the platforms our binary packages run on. We can remain forwards compatible from 10.7 up if we pick either of the latter two choices as explained in the [SDK docs]( https://developer.apple.com/library/mac/documentation/DeveloperTools/Conceptual/cross_development/Overview/overview.html ) (may need to login). Please share your thoughts.

cc @pelson @ocefpaf @msarahan @jjhelmus

**Comment by pelson** [comment]
> Do we just accept this and change the toolchain to match this constraint? 

Given we don't provide linux-32 builds simply because the effort outweighs the benefit, so too I imagine the situation on OSX. Does 10.9 have a reasonable market share? Do our existing binaries work on <10.9? (if we can't find anyone to verify, we aren't going to find anyone to defend the legacy side either)

Let's go with the simple solution, and we can iterate if there is sufficient groundswell (and in a perfect world, some help from those that do want it).

**Comment by patricksnape** [comment]
I agree that targeting 10.9 is probably the pragmatic and sensible solution for the time being.

**Comment by jakirkham** [comment]
So, I'm still using 10.9 and there are some lab computers that are still on 10.9. Though I think we are going to upgrade them. I can confirm that the binaries built on conda-forge work just fine for me on 10.9.

As for market share, this discussion came up over at NumPy before and one piece of data that was used was where Adium was updated. This isn't the most representative thing (at least as was determined from the discussion at NumPy) as the feeling was that people were slow to upgrade OSes after getting a working build of the SciPy stack over concerns of breaks and the need to rebuild. I would hope that affects `conda` users less as this is suppose to make it easy to get packages built and installed. Especially, I would hope that conda-forge makes this problem nearly non-existent as the onus of building on one's own machine is basically removed except for hardcore debugging.

In any event, the Adium [download counts]( https://www.adium.im/sparkle/#osVersion ) suggest that 10.6 is still a thing (though quite small), but that ends up being painfully hard to support (as it had universal builds and could be installed on 32-bit machines) and I would really recommend that we let that ship sail especially as Continuum has (I say that even though I have an old 10.6 MacBook Pro lying around). After that it appears all significant numbers are for 10.9 and higher. If one looks at the [detailed download counts]( https://www.adium.im/sparkle/?forceShow[]=osVersion#osVersion ), one can see some truly ancient OSes in use like 10.4 and earlier these are probably still in use as they supported PPC so may not even be Intel machines. Beyond these very strange cases there is some use of 10.8. Though I think those users are probably just being lazy and not upgrading as 10.9 and up was offered for free to them and I don't believe there was anything special dropped by going through that upgrade path.

In short, supporting 10.9 is totally sensible to me and we actually are doing that now. The only question (at least in my mind) is do we want to make that explicit. The concern here being making it explicit stops people from building 10.7 binaries with our recipes. Currently, without making any changes here, one can build 10.7 supported binaries if they have the right SDK. This matters for Continuum as they do advertise 10.7 support. Though IIRC they rebuild everything from source anyways (unless that has recently changed). In any event, I would like to keep the building from source option open to them if possible. It would be nice if @msarahan could comment on this.

**Comment by patricksnape** [comment]
We could always ask Travis if they'd be willing to update their OSX image to include the 10.7 SDK. We have it installed on our OSX jenkins instance running 10.11 and it seems to work fine just having the SDK unpacked in the correct directory.

**Comment by jakirkham** [comment]
I'm not suggesting that we go down that road. Merely stating that we may want to the leave the door open for those who wish to build for 10.7 themselves.

**Comment by pelson** [comment]
I'd sooner not close any doors, but equally don't think we need to do anything at this moment beyond using the 10.9 provided by TravisCI. If the issue rises its head in a compelling way, I believe we have good options to broaden support to earlier versions.

**Comment by pelson** [comment]
Sounds reasonable to me. We currently don't have any compatibility type docs, but certainly should have. @jakirkham - any suggestions on the best place for it?

**Comment by jakirkham** [comment]
Just for my edification, @patricksnape, what version of XCode are you using with Jenkins where you are installing the 10.7 SDK?

Agreed, that this should live in the docs. Adding another file to the [main docs]( https://github.com/conda-forge/conda-forge.github.io/tree/master/docs ) explaining how our CIs are configured seems wise. That should make it pretty visible. ( http://conda-forge.rtfd.io/ )

**Comment by patricksnape** [comment]
`xcodebuild -version` reports `Xcode 6.4 Build version 6E35b`

**Comment by 183amir** [comment]
I vote for supporting `10.9` and above too. I think **even if** this might be a bad decision now, it will be a reasonable decision in the future :smile:.

**Comment by jakirkham** [comment]
To summarize the problem that gave rise to this change so as to avoid [future confusion]( http://xkcd.com/979/ ).

![wisdom_of_the_ancients](https://cloud.githubusercontent.com/assets/3019665/19706692/48d80624-9ae1-11e6-9788-cb5627d0e73b.png)

One should be able to set `MACOSX_DEPLOYMENT_TARGET`, in theory, and get something built for the right target Mac OS version. However, in practice, signs point to this not being the case. Please see this [build]( https://travis-ci.org/conda-forge/staged-recipes/jobs/133821737#L519-L531 ) where `cmake` decides to use the 10.9 SDK even though the 10.7 SDK is selected because the 10.7 SDK is not provided. I find this still to be a relevant lesson even today.

cc @conda-forge/core @mingwandroid
